Finite lorax support

quic-jouachen opened the pull request:
- Use case: Users can activate multiple LoRA adapters and compile them with base model. At runtime, they can specify which prompt should utilize which adapter, allowing for mixed adapter usage within the same batch. (refer to examples/lora_models.py for more details)
- These changes are built for both continuous batching and regular inference scenario.

quic-mohitgupta commented on QEfficient/exporter/export_hf_to_cloud_ai_100.py:
```diff
@@ -14,7 +14,7 @@
 
 import QEfficient
-from QEfficient.base.common import AUTO_MODEL_MAP_TO_MODEL_TYPE_MAP, QEFF_MODEL_TYPE, QEFFCommonLoader
+from QEfficient.base.common import QEFFCommonLoader
```
Restore change to  AUTO_MODEL_MAP_TO_MODEL_TYPE_MAP, QEFF_MODEL_TYPE

quic-jouachen replied:
reverted.

quic-mohitgupta commented on QEfficient/exporter/export_hf_to_cloud_ai_100.py:
```diff
@@ -326,19 +330,20 @@ def export_for_cloud(
         )
 
     # FIXME: move all this to class instead of here, and just call qeff_model.export here.
-    if AUTO_MODEL_MAP_TO_MODEL_TYPE_MAP.get(qeff_model.__class__, None) == QEFF_MODEL_TYPE.CAUSALLM:  # type: ignore
```
Restore the "if" statement

quic-jouachen replied:
reverted

quic-mohitgupta commented on QEfficient/exporter/export_utils.py:
```diff
@@ -83,6 +83,8 @@ def export_onnx(
         elif iname == "batch_index":
             dynamic_axes[iname] = {0: "batch_size"}
+        elif iname == "lora_ids":
+            dynamic_axes[iname] = {0: "batch_size"}
```
what is default values for lora_ids? I should be INT_MAX

quic-jouachen replied:
Have tried initializing them as INT_MAX but the CtxGatherFuncCB pytorch script will show out of index errors. Therefore, I keep it as zero for now.

quic-mohitgupta replied:
If we can not do it compile time/ default to INT_MAX at run time.

quic-jouachen replied:
added in QEffAutoLoraModelForCausalLM.run_cloud_ai_100()

quic-mohitgupta commented on QEfficient/lora/auto.py:
```diff
@@ -0,0 +1,396 @@
+                        target_module == "q_proj"
+                        or target_module == "k_proj"
+                        or target_module == "v_proj"
+                        or target_module == "up_proj"
```
remove "up_proj"

quic-jouachen replied:
removed

quic-mohitgupta commented on QEfficient/lora/auto.py:
```diff
@@ -0,0 +1,396 @@
+                                f"base_model.model.model.layers.{i}.self_attn.{target_module}.lora_B.weight"
+                            ]
+                        )
+                    elif target_module == "up_proj" or target_module == "gate_proj" or target_module == "down_proj":
```
remove this elif for up_proj/gate_proj/down_proj

quic-jouachen replied:
removed

quic-mohitgupta commented on QEfficient/lora/auto.py:
```diff
@@ -0,0 +1,396 @@
+                    module = self.model.model.layers[i].self_attn.k_proj
+                elif target_module == "v_proj":
+                    module = self.model.model.layers[i].self_attn.v_proj
+                elif target_module == "o_proj":
```
remove all down modules.

quic-jouachen replied:
removed

quic-mohitgupta commented on QEfficient/lora/auto.py:
```diff
@@ -0,0 +1,396 @@
+        else:
+            export_dir = Path(export_dir).with_name(export_dir.name + "-" + self.model_hash)
+        onnx_dir_path = os.path.join(export_dir, "onnx")
+        model_base_name = self.model_card_name.replace("/", "_") + "_kv"
```
Instead of adding "kv" we should add "lora" in model base name.

quic-mohitgupta commented on QEfficient/lora/auto.py:
```diff
@@ -0,0 +1,396 @@
+        # Export
+        os.makedirs(onnx_dir_path, exist_ok=True)
+        _, onnx_model_path = QEfficient.export(
+            model_name=self.model_card_name,
```
model_name should be model_base_name

quic-jouachen replied:
This is to align with the first onnx name generated by QEfficient.export(), so will keep it

quic-mohitgupta commented on tests/lora/test_lora_model.py:
```diff
@@ -0,0 +1,198 @@
+    assert len(qeff_model.active_adapter_to_id) == 0
+
+
+def test_auto_lora_model_for_causal_lm_hash():
```
add test when hash is equal.

quic-jouachen replied:
added

quic-mohitgupta commented on tests/lora/test_lora_model.py:
```diff
@@ -0,0 +1,198 @@
+
+
+@pytest.mark.parametrize("base_model_name,adapter_id_0,adapter_id_1", model_samples[:1])
+def test_auto_lora_model_for_causal_lm_load_set_adapter(base_model_name, adapter_id_0, adapter_id_1):
```
add some description and add id_check in the name.

quic-jouachen replied:
added

quic-mohitgupta commented on QEfficient/lora/layers.py:
```diff
@@ -0,0 +1,58 @@
+import math
```
Missing license

quic-jouachen replied:
added

quic-rishinr commented on QEfficient/generation/text_generation_inference.py:
```diff
@@ -460,6 +470,10 @@ def prepare_decode_inputs(self):
         if self.batch_index is not None:
             decode_inputs["batch_index"] = self.batch_index
 
+        if self.prompt_to_lora_id_mapping_decode and self.full_batch_size is not None:
```
is it possible to run lorax in non-CB mode?

quic-jouachen replied:
It only support CB for now as the export and compile is implement for CB only.

quic-jouachen replied:
Will have to test it and update the code.

quic-rishinr commented:
@irajagop @ochougul please review

quic-jouachen commented on QEfficient/generation/text_generation_inference.py:
```diff
@@ -634,6 +653,12 @@ def run_continuous_batching_decode(self, prompt_queue, generation_len):
                     )
 
                     generated_id_current_index[decode_batch_id] += 1
+
```
This is to ensure the next decode session run gets correct lora_ids as input. I have observed if putting this inside else condition, last (few) token output won't match the baseline. And thus, making the change.

quic-rishinr replied:
if its outside the loop the decode_inputs["lora_ids"] will get updated every time. I believe we need to update it only if we are running prefill when one batch_id is complete. Can you try adding the condition inside if prompt_queue condition?

quic-jouachen replied:
Got your point, just updated a patch based on the suggestion. Thanks!

quic-jouachen commented on QEfficient/lora/auto.py:
```diff
@@ -0,0 +1,428 @@
+
+class QEffAutoLoraModelForCausalLM(QEFFAutoModelForCausalLM):
+    """
+    QEff class for loading models with mutltiple LoRA adapters.
```
fixed

quic-jouachen commented on QEfficient/lora/auto.py:
```diff
@@ -0,0 +1,428 @@
+    def unload_adapter(self, adapter_name: str):
+        # remove from active list
+        if adapter_name not in self.active_adapters:
+            print(f"Adapter name {adapter_name} is not set active yet")
```
fixed

quic-jouachen commented on QEfficient/lora/auto.py:
```diff
@@ -0,0 +1,428 @@
+                c_tensor_list = list(range(self.max_num_adapters + 1))
+
+                for lora_name, lora_id in self.active_adapter_to_id.items():
+                    if (
```
That looks better, thanks, have fixed

quic-jouachen commented on QEfficient/lora/layers.py:
```diff
@@ -0,0 +1,58 @@
+
+class LinearMultiLoRA(nn.Linear):
+    def multilora_init(self, lora_rank, max_num_adapters):
+        self.max_num_adapters = max_num_adapters
```
added

quic-jouachen commented on QEfficient/lora/auto.py:
```diff
@@ -0,0 +1,428 @@
+            self.adapter_configs[adapter_name] = PeftConfig.from_pretrained(adapter_model_id)
+
+        # check if adapters has same target module and rank
+        assert (
```
Fixed

quic-jouachen commented on QEfficient/lora/auto.py:
```diff
@@ -0,0 +1,428 @@
+        "Initialize the fixed lora model with multiple adapter weigths standby"
+
+        # assume all adapters have same target_modules and ranks
+        assert self.max_num_adapters == len(self.active_adapters), "Inconsistent max_num_adapters and active_adapters"
```
fixed

quic-jouachen commented on QEfficient/lora/auto.py:
```diff
@@ -0,0 +1,428 @@
+            :adapter_name (str): Adapter name to be used to set this adapter as current
+        """
+        if (adapter_name in self.adapter_weights.keys()) and (adapter_name in self.adapter_configs.keys()):
+            logger.warning(f"Overwrite weights and configs for adapter name {adapter_name}")
```
we will change to warning user if adapter_name is already use (instead of overwriting)

quic-jouachen commented on QEfficient/lora/auto.py:
```diff
@@ -0,0 +1,428 @@
+        self.adapter_weights[adapter_name] = {
+            k: v.numpy().astype("float16") for k, v in load_peft_weights(adapter_model_id).items()
+        }
+        self.adapter_configs[adapter_name] = PeftConfig.from_pretrained(adapter_model_id)
```
Added conditional check to avoid re-download.

quic-jouachen commented on QEfficient/lora/auto.py:
```diff
@@ -0,0 +1,428 @@
+            self.adapter_weights[adapter_name] = adapter_weight
+            self.adapter_configs[adapter_name] = adapter_config
+        else:  # load from hugging face
+            self.adapter_weights[adapter_name] = {
```
We have changed to reuse this API inside load_adapter(). download_adapter() is still needed as a standalone API to support scenario that caches multiple adapters and activate them later

quic-jouachen commented on QEfficient/lora/auto.py:
```diff
@@ -0,0 +1,428 @@
+        }
+        self.adapter_configs[adapter_name] = PeftConfig.from_pretrained(adapter_model_id)
+
+    def load_adapter(self, adapter_model_id: str, adapter_name: str, **kwargs: Any):
```
Fixed

quic-jouachen commented on QEfficient/lora/layers.py:
```diff
@@ -0,0 +1,58 @@
+        self.max_num_adapters = max_num_adapters
+        self.lora_rank = lora_rank
+
+        self.lora_weight_A = nn.Parameter(
```
Has renamed to 

- lora_weight_A -> lora_a_weights
- lora_weight_B -> lora_b_weights
- lora_weight_C  -> lora_scalings

Will detailed more in docstring to explain

quic-jouachen commented on QEfficient/lora/auto.py:
```diff
@@ -0,0 +1,428 @@
+        # load_weight to model
+        self.load_adapter_weights_to_model()
+
+    def export(self, **kwargs) -> str:
```
We have prepared a local patch to address this. Will update once PR#116 is merged.

ochougul reviewed (changes requested):
I looked at the way transformers and peft packages handle loading and doing inference with adapters.
* All the methods for loading and running inference with adapters are implemented in `PeftModel` which is inherited by `AutoPeftModel`. Instance of `AutoPeftModel`  is returned by `AutoPeftModelForCausalLM.from_pretrained`
* transformers `AutoModelForCausalLM` inherits `PeftAdapterMixin` that provides the same methods and in turn uses low level APIs from peft package to provide the simlar utility of loading adapter and running inference.

I don't think there is a need to make a new `QEffAutoLoraModelForCausalLM` class, that is not known to users as most of the users will be using `peft`, `transformers` packages in their day-to-day job.

It would be best to keep the interface same as `peft`, `transformers` packages unless there is an absolute need of new interface. Let's have a discussion if needed.

quic-jouachen commented:
> I looked at the way transformers and peft packages handle loading and doing inference with adapters.
> 
> * All the methods for loading and running inference with adapters are implemented in `PeftModel` which is inherited by `AutoPeftModel`. Instance of `AutoPeftModel`  is returned by `AutoPeftModelForCausalLM.from_pretrained`
> * transformers `AutoModelForCausalLM` inherits `PeftAdapterMixin` that provides the same methods and in turn uses low level APIs from peft package to provide the simlar utility of loading adapter and running inference.
> 
> I don't think there is a need to make a new `QEffAutoLoraModelForCausalLM` class, that is not known to users as most of the users will be using `peft`, `transformers` packages in their day-to-day job.
> 
> It would be best to keep the interface same as `peft`, `transformers` packages unless there is an absolute need of new interface. Let's have a discussion if needed.

Thanks for the suggestion, Onkar! We had discussions with Anuj and IIlango on this and went over the design. This is a different usecase then peft, Here we are doing finite lora which need lora_mask to be passed in (by which user can specify different lora id for different batch element) and also doing custom Linear lora layer implementation optimized for our hardware.

ochougul commented on QEfficient/generation/text_generation_inference.py:
```diff
@@ -625,6 +657,12 @@ def run_continuous_batching_decode(self, prompt_queue, generation_len):
 
                         self.session.set_buffers({"logits": logits_out_placeholder})
                         decode_pause_time += perf_counter() - start
+
+                        if self.prompt_to_lora_id_mapping_decode:
+                            decode_inputs["lora_ids"][decode_batch_id] = self.prompt_to_lora_id_mapping_decode[
+                                batch_id_map[decode_batch_id]
+                            ]
+
```
can this go inside `_update_decode_input` method?

quic-jouachen replied:
Due to the necessity of referring to the correct id from `batch_id_map `(which is not in `_update_decode_input` but here), I need to include this line here.

ochougul reviewed (changes requested):
review is still WIP

quic-rishinr commented on QEfficient/lora/auto.py:
```diff
@@ -0,0 +1,366 @@
+            logger.info(f"Adapter name {adapter_name} is not set active yet")
+            return False
+
+        self.max_num_adapters -= 1
```
Do we really need self.max_num_adapters variable? the number of active adapters can be always derived from len(self.active_adapter_to_id)

quic-jouachen replied:
I have removed the `self.max_num_adapters` and use `len(self.active_adapter_to_id)` every time needed.

quic-rishinr commented on QEfficient/lora/auto.py:
```diff
@@ -0,0 +1,366 @@
+        """
+
+        # initialize the adapter model
+        assert self.max_num_adapters, "Please use load_adapter() to add at least one adapter; otherwise, refer to QEFFAutoModelForCausalLM for base model usage"
```
please raise a ValueError instead of assert.

quic-jouachen replied:
Updated.

quic-jouachen commented on QEfficient/lora/auto.py:
```diff
@@ -0,0 +1,366 @@
+        prompts: List[str],
+        device_id: List[int] = None,
+        runtime: str = "AI_100",
+        **kwargs,
```
I believe this scenario is already addressed in `run_continuous_batching_decode()` using `prompt_queue` in `text_generation_inference.py`, even for the base model. Please let me know if I’ve overlooked anything.

quic-jouachen commented on QEfficient/generation/text_generation_inference.py:
```diff
@@ -342,6 +345,16 @@ def __init__(
             full_batch_size if full_batch_size else self._fetch_full_batch_size()
         )  # Check and fetch full batch size if CB is enabled
 
+        if prompt_to_lora_id_mapping:
```
Thanks for the suggestion. I added this line to the `generate()` function in `lora/auto.py` so that the user can directly modifies the list they passed in. I set `prompt_to_adapter_mapping` to default to ‘base’ if not provided. This way, `lora_id` will default to 0 and use the base model.

quic-jouachen commented on QEfficient/lora/auto.py:
```diff
@@ -0,0 +1,366 @@
+            :device_id (List[int]): Ids of devices for running the qpc pass as [0] in case of normal model / [0, 1, 2, 3] in case of tensor slicing model
+        ``optional`` Args:
+            :runtime (str, optional): Only ``AI_100`` runtime is supported as of now; ``ONNXRT`` and ``PyTorch`` coming soon. Defaults to "AI_100".
+            :prompt_to_adapter_mapping (List[str]): A list of adapter names that maps to the prompts, specifying which adapter the prompt wants to apply. "base" for base model (no adapter).
```
I have added these lines in the doc string.

quic-jouachen commented on QEfficient/lora/auto.py:
```diff
@@ -0,0 +1,366 @@
+            :device_id (List[int]): Ids of devices for running the qpc pass as [0] in case of normal model / [0, 1, 2, 3] in case of tensor slicing model
+        ``optional`` Args:
+            :runtime (str, optional): Only ``AI_100`` runtime is supported as of now; ``ONNXRT`` and ``PyTorch`` coming soon. Defaults to "AI_100".
+            :prompt_to_adapter_mapping (List[str]): A list of adapter names that maps to the prompts, specifying which adapter the prompt wants to apply. "base" for base model (no adapter).
```
I have marked it as mandatory. I also had set `prompt_to_adapter_mapping` to default to ‘base’ if not provided. This way, `lora_id` will default to 0 and use the base model.

quic-jouachen commented on QEfficient/generation/text_generation_inference.py:
```diff
@@ -461,6 +474,16 @@ def prepare_decode_inputs(self):
 
+        if self.prompt_to_lora_id_mapping_decode:
+            if self.full_batch_size:
+                first_batch_lora_ids = [self.prompt_to_lora_id_mapping_decode[i] for i in range(self.full_batch_size)]
```
We do not remove any entries from `prompt_to_lora_id_mapping_decode` in the continuous batching scenario, as the decode phase of each prompt might overlap and reference the `lora_id` again later.

quic-jouachen commented on QEfficient/lora/auto.py:
```diff
@@ -0,0 +1,366 @@
+
+        Args:
+            :adapter_model_id (str): Adapter model ID from huggingface hub or local path
+            :adapter_name (str): Adapter name to be used to set this adapter as current
```
updated.

ochougul commented on QEfficient/peft/auto.py:
```diff
@@ -38,6 +39,7 @@ class QEffAutoPeftModelForCausalLM(QEFFBaseModel):
 
     Args:
         :model (nn.Module): PyTorch model
+        :finite_adapters (bool): set True to enable finite adapter mode with QEffAutoLoraModelForCausalLM class. Please refer to QEffAutoLoraModelForCausalLM for API specification.
```
add this in `from_pretrained` as well.

ochougul commented on tests/lora/test_lora_model.py:
```diff
@@ -0,0 +1,218 @@
+def test_auto_peft_model_for_causal_lm_from_pretrained(base_model_name, adapter_id_0, adapter_id_1):
+    qeff_model = QEffAutoPeftModelForCausalLM.from_pretrained(adapter_id_0, "id_0", finite_adapters=True)
+
+    assert qeff_model.base_model_name == base_model_name
```
add 
```
assert isinstance(qeff_model, QEffAutoLoraModelForCausalLM)
```

ochougul reviewed (changes requested):
Can you please move lora module inside peft module so the directory structure can be
```
peft
|__lora
```

quic-jouachen commented on QEfficient/lora/auto.py:
```diff
@@ -0,0 +1,384 @@
+            type(self.model).__name__ == "QEffMistralForCausalLM" or type(self.model).__name__ == "QEffLlamaForCausalLM"
+        ), f"Only QEffMistralForCausalLM and QEffLlamaForCausalLM model are supported but get {type(self.model).__name__}"
+
+        self.base_model_name = self.model.model.config._name_or_path
```
Removed

quic-jouachen commented:
> raise `TypeError("required adapter name argument")`

I've added a check in the `QEffAutoPeftModelForCausalLM` to see if adapter_name is passed in and if its string; otherwise, will raise type error. I also added additional check in the unit test.

ochougul reviewed (approved):
LGTM
